File: src/factories/internal/metadata/iterate_metadata_sort.ts
import { Metadata } from "../../../schemas/metadata/Metadata";
import { MetadataObjectType } from "../../../schemas/metadata/MetadataObjectType";

import { MetadataCollection } from "../../MetadataCollection";

export const iterate_metadata_sort = (props: {
  collection: MetadataCollection;
  metadata: Metadata;
}) => {
  const visited: Set<Metadata> = new Set();
  const next = (metadata: Metadata) =>
    iterate({
      visited,
      collection: props.collection,
      metadata,
    });
  for (const array of props.collection.arrays()) next(array.value);
  for (const tuple of props.collection.tuples())
    for (const element of tuple.elements) next(element);
  for (const object of props.collection.objects())
    for (const property of object.properties) next(property.value);
  next(props.metadata);
};

const iterate = (props: {
  visited: Set<Metadata>;
  collection: MetadataCollection;
  metadata: Metadata;
}) => {
  if (props.visited.has(props.metadata)) return;
  props.visited.add(props.metadata);

  const next = (metadata: Metadata) =>
    iterate({
      ...props,
      metadata,
    });

  // ITERATE CHILDREN
  for (const map of props.metadata.maps) next(map.value);
  for (const set of props.metadata.sets) next(set.value);
  if (props.metadata.escaped !== null) next(props.metadata.escaped.returns);
  if (props.metadata.rest !== null) next(props.metadata.rest);

  // SORT OBJECTS
  if (props.metadata.objects.length > 1) {
    props.metadata.objects.sort((x, y) =>
      MetadataObjectType.covers(x.type, y.type)
        ? -1
        : MetadataObjectType.covers(y.type, x.type)
          ? 1
          : 0,
    );
    props.metadata.union_index = props.collection.getUnionIndex(props.metadata);
  }

  // SORT ARRAYS AND TUPLES
  if (props.metadata.arrays.length > 1)
    props.metadata.arrays.sort((x, y) =>
      Metadata.covers(x.type.value, y.type.value)
        ? -1
        : Metadata.covers(y.type.value, x.type.value)
          ? 1
          : 0,
    );
  if (props.metadata.tuples.length > 1)
    props.metadata.tuples.sort((x, y) => {
      const xt = Metadata.initialize();
      const yt = Metadata.initialize();

      xt.tuples.push(x);
      yt.tuples.push(y);

      return Metadata.covers(xt, yt) ? -1 : Metadata.covers(yt, xt) ? 1 : 0;
    });

  // SORT CONSTANT VALUES
  for (const constant of props.metadata.constants)
    if (constant.type === "string") constant.values.sort();
    else if (constant.type === "number")
      constant.values.sort((a, b) => (a.value as number) - (b.value as number));
    else if (constant.type === "bigint")
      constant.values.sort((a, b) =>
        (a.value as bigint) < (b.value as bigint) ? -1 : 1,
      );
    else constant.values.sort((a, _b) => (a.value === false ? -1 : 1));
};

File: packages/typia/native/core/factories/internal/metadata/iterate_metadata_sort.go
package metadata

import (
  "fmt"
  "slices"
  "sort"
  "strings"

  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

func Iterate_metadata_sort(props struct {
  Collection *schemametadata.MetadataCollection
  Metadata   *schemametadata.MetadataSchema
}) {
  visited := map[*schemametadata.MetadataSchema]bool{}
  objectKeys := map[*schemametadata.MetadataObjectType]iterate_metadata_sort_object_key{}
  next := func(metadata *schemametadata.MetadataSchema) {}
  next = func(metadata *schemametadata.MetadataSchema) {
    iterate_metadata_sort_iterate(struct {
      visited    map[*schemametadata.MetadataSchema]bool
      collection *schemametadata.MetadataCollection
      metadata   *schemametadata.MetadataSchema
      next       func(*schemametadata.MetadataSchema)
      objectKeys map[*schemametadata.MetadataObjectType]iterate_metadata_sort_object_key
    }{
      visited:    visited,
      collection: props.Collection,
      metadata:   metadata,
      next:       next,
      objectKeys: objectKeys,
    })
  }
  for _, array := range props.Collection.Arrays() {
    next(array.Value)
  }
  for _, tuple := range props.Collection.Tuples() {
    for _, element := range tuple.Elements {
      next(element)
    }
  }
  for _, object := range props.Collection.Objects() {
    for _, property := range object.Properties {
      next(property.Value)
    }
  }
  next(props.Metadata)
}

func iterate_metadata_sort_iterate(props struct {
  visited    map[*schemametadata.MetadataSchema]bool
  collection *schemametadata.MetadataCollection
  metadata   *schemametadata.MetadataSchema
  next       func(*schemametadata.MetadataSchema)
  objectKeys map[*schemametadata.MetadataObjectType]iterate_metadata_sort_object_key
}) {
  if props.visited[props.metadata] {
    return
  }
  props.visited[props.metadata] = true

  for _, item := range props.metadata.Maps {
    props.next(item.Value)
  }
  for _, item := range props.metadata.Sets {
    props.next(item.Value)
  }
  if props.metadata.Escaped != nil {
    props.next(props.metadata.Escaped.Returns)
  }
  if props.metadata.Rest != nil {
    props.next(props.metadata.Rest)
  }

  if len(props.metadata.Objects) > 1 {
    sort.SliceStable(props.metadata.Objects, func(i, j int) bool {
      x := props.metadata.Objects[i]
      y := props.metadata.Objects[j]
      if iterate_metadata_sort_object_covers(props.objectKeys, x.Type, y.Type) {
        return true
      }
      if iterate_metadata_sort_object_covers(props.objectKeys, y.Type, x.Type) {
        return false
      }
      return false
    })
    index := props.collection.GetUnionIndex(props.metadata)
    props.metadata.Union_index = &index
  }

  if len(props.metadata.Arrays) > 1 {
    sort.SliceStable(props.metadata.Arrays, func(i, j int) bool {
      x := props.metadata.Arrays[i]
      y := props.metadata.Arrays[j]
      if schemametadata.MetadataSchema_covers(x.Type.Value, y.Type.Value) {
        return true
      }
      if schemametadata.MetadataSchema_covers(y.Type.Value, x.Type.Value) {
        return false
      }
      return false
    })
  }
  if len(props.metadata.Tuples) > 1 {
    sort.SliceStable(props.metadata.Tuples, func(i, j int) bool {
      xt := schemametadata.MetadataSchema_initialize()
      yt := schemametadata.MetadataSchema_initialize()
      xt.Tuples = append(xt.Tuples, props.metadata.Tuples[i])
      yt.Tuples = append(yt.Tuples, props.metadata.Tuples[j])
      if schemametadata.MetadataSchema_covers(xt, yt) {
        return true
      }
      if schemametadata.MetadataSchema_covers(yt, xt) {
        return false
      }
      return false
    })
  }

  for _, constant := range props.metadata.Constants {
    if constant.Type == "string" {
      sort.SliceStable(constant.Values, func(i, j int) bool {
        return fmt.Sprint(constant.Values[i].Value) < fmt.Sprint(constant.Values[j].Value)
      })
    } else if constant.Type == "number" {
      sort.SliceStable(constant.Values, func(i, j int) bool {
        return iterate_metadata_sort_float(constant.Values[i].Value) < iterate_metadata_sort_float(constant.Values[j].Value)
      })
    } else if constant.Type == "bigint" {
      sort.SliceStable(constant.Values, func(i, j int) bool {
        return fmt.Sprint(constant.Values[i].Value) < fmt.Sprint(constant.Values[j].Value)
      })
    } else {
      sort.SliceStable(constant.Values, func(i, j int) bool {
        value, _ := constant.Values[i].Value.(bool)
        return value == false
      })
    }
  }
}

func iterate_metadata_sort_object_covers(
  cache map[*schemametadata.MetadataObjectType]iterate_metadata_sort_object_key,
  x *schemametadata.MetadataObjectType,
  y *schemametadata.MetadataObjectType,
) bool {
  if x == nil || y == nil || len(x.Properties) < len(y.Properties) {
    return false
  }
  return iterate_metadata_sort_object_keys(cache, x).Signature ==
    iterate_metadata_sort_object_keys(cache, y).Signature
}

type iterate_metadata_sort_object_key struct {
  Signature string
}

func iterate_metadata_sort_object_keys(
  cache map[*schemametadata.MetadataObjectType]iterate_metadata_sort_object_key,
  object *schemametadata.MetadataObjectType,
) iterate_metadata_sort_object_key {
  if value, ok := cache[object]; ok {
    return value
  }
  keys := []string{}
  if object != nil {
    for _, property := range object.Properties {
      if property != nil && property.Key != nil {
        keys = append(keys, property.Key.GetName())
      }
    }
  }
  slices.Sort(keys)
  value := iterate_metadata_sort_object_key{
    Signature: strings.Join(keys, "\x00"),
  }
  cache[object] = value
  return value
}

func iterate_metadata_sort_float(value any) float64 {
  switch v := value.(type) {
  case float64:
    return v
  case float32:
    return float64(v)
  case int:
    return float64(v)
  case int64:
    return float64(v)
  case int32:
    return float64(v)
  case uint:
    return float64(v)
  case uint64:
    return float64(v)
  case uint32:
    return float64(v)
  default:
    return 0
  }
}
